### PR TITLE
CAL-16 STANAG 4559 endpoint that supports the ProductMgr interface

### PR DESCRIPTION
**Pull request by troymohl**
#### What does this PR do?
Adds the ProductMgr interface implementation for the NSILI Endpoint.
Increases JUnit test coverage of the NSILI endpoint classes.
#### Who is reviewing it?
@jaymcnallie @bdeining @dcruver 
#### How should this be tested?
Run JUnit tests
#### What are the relevant tickets?
https://codice.atlassian.net/browse/CAL-16
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 1]
```diff
@@ -89,12 +141,31 @@
 
     private static DataModelMgr dataModelMgr;
 
+    private static StandingQueryMgr standingQueryMgr;
+
+    //    private static final String ENCODING = "UTF-8";
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove?

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 2]
```diff
@@ -276,63 +330,299 @@ public void retrieveProductFromDAG(DAG dag) throws MalformedURLException {
         }
     }
 
-    public void validate_order(ORB orb) throws Exception {
+    //    public void validate_order(ORB orb) throws Exception {
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 3]
```diff
@@ -276,63 +330,299 @@ public void retrieveProductFromDAG(DAG dag) throws MalformedURLException {
...
+
+                org.omg.CORBA.Object obj = poa.create_reference_with_id(callbackId.getBytes(Charset.forName(ENCODING)),
+                        CallbackHelper.id());
+
+                Callback callback = CallbackHelper.narrow(obj);
+
+                String standingQueryCallbackId = standingQueryRequest.register_callback(callback);
+                nsiliCallback.setCallbackID(standingQueryCallbackId);
+
+                //            System.out.println("Checking if results are already available.");
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 4]
```diff
@@ -450,6 +766,91 @@ public String getIorTextFile(String iorURL) throws Exception {
...
+        }
+    }
+
+    public void testCallbackCatalogMgr(ORB orb, POA poa, Query query) throws Exception {
+        if (catalogMgr != null) {
+            System.out.println("Testing Query Results via Callback ...");
+            SortAttribute[] sortAttributes = getSortableAttributes();
+            String[] resultAttributes = getResultAttributes();
+
+            //            TestNsiliCallback nsiliCallback = new TestNsiliCallback();
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 5]
```diff
@@ -450,6 +766,91 @@ public String getIorTextFile(String iorURL) throws Exception {
...
+            String catalogSearchCallbackID = catalogSearchQueryRequest.register_callback(
+                    CallbackHelper.narrow(obj));
+            nsiliCallback.setCallbackID(catalogSearchCallbackID);
+            callbacks.add(nsiliCallback);
+
+            System.out.println(
+                    "Callback Catalog Mgr Callback registered: " + catalogSearchCallbackID);
+
+            //            DAGListHolder dagListHolder = new DAGListHolder();
+            //            catalogSearchQueryRequest.complete_DAG_results(dagListHolder);
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 6]
```diff
@@ -450,6 +766,91 @@ public String getIorTextFile(String iorURL) throws Exception {
...
+    }
+
+    private void deregisterStandingQueryCallbacks()
+            throws InvalidInputParameter, SystemFault, ProcessingFault, ObjectNotActive,
+            WrongPolicy {
+        for (TestNsiliCallback callback : callbacks) {
+            System.out.println("Freeing callback: " + callback.getCallbackID());
+            callback.getQueryRequest()
+                    .free_callback(callback.getCallbackID());
+            //            poa.deactivate_object(callback._object_id());
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 7]
```diff
@@ -479,4 +880,240 @@ public void checkServerTrusted(X509Certificate[] x509Certificates, String s)
...
+
+        @Override
+        public void _notify(State theState, RequestDescription description)
+                throws InvalidInputParameter, ProcessingFault, SystemFault {
+            System.out.println("**************************************************************");
+            System.out.println("******************* NOTIFY CALLED ****************************");
+            System.out.println("**************************************************************");
+            try {
+                System.out.println("--------  TestNsiliCallback.notify --------");
+                //                System.out.println("State: " + theState);
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java [hunk 8]
```diff
@@ -479,4 +880,240 @@ public void checkServerTrusted(X509Certificate[] x509Certificates, String s)
...
+                //                            }
+                //                        }
+                //                    }
+                //                } else {
+                //                    System.out.println("Notified with no details");
+                //                }
+                //                }
+
+                System.out.println("Results from notification: ");
+                //                DAGListHolder dagListHolder = new DAGListHolder();
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/NsiliClient.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/Client.java [hunk 9]
```diff
@@ -42,40 +66,96 @@ public static void main(String args[]) throws Exception {
...
+            System.out.println("Unable to obtain IOR File :  Must specify URL to IOR file.");
+        }
+        client.runTest(args);
+    }
+
+    private void startHttpListener() {
+        try {
+            ResourceConfig rc = new PackagesResourceConfig(
+                    "com.connexta.alliance.nsili.mockserver.client");
+//            SSLContextConfigurator sslCon = new SSLContextConfigurator();
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/Client.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/Client.java [hunk 10]
```diff
@@ -42,40 +66,96 @@ public static void main(String args[]) throws Exception {
...
+                //ProductMgr
+                System.out.println("-----------------------");
+                try {
+                    nsiliClient.testProductMgr(orb, rootPOA, results);
+                } catch (Exception e) {
+                    System.out.println("Unable to test ProductMgr: " + NsilCorbaExceptionUtil.getExceptionDetails(e));
+                }
+                System.out.println("-----------------------");
+
+                // OrderMgr
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/Client.java]: Remove commented code

File: distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/Client.java [hunk 11]
```diff
@@ -42,40 +66,96 @@ public static void main(String args[]) throws Exception {
         String[] managers = nsiliClient.getManagerTypes();
         nsiliClient.initManagers(managers);
 
+        // Standing Query Mgr
+        Query standingAllQuery = new Query(NsiliConstants.NSIL_ALL_VIEW, "NSIL_CARD.identifier like '%'");
+//        nsiliClient.testStandingQueryMgr(orb, rootPOA, standingAllQuery);
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/com/connexta/alliance/nsili/mockserver/client/Client.java]: Remove commented code

File: catalog/nsili/catalog-nsili-endpoint/src/test/java/com/connexta/alliance/nsili/endpoint/TestGetRelatedFilesRequestImpl.java [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+        assertThat(locations.value, notNullValue());
+        assertThat(locations.value.length, is(1));
+    }
+
+    @Test
+    public void testComplete202() throws Exception {
+        when(mockHttpClient.execute(any(HttpUriRequest.class))).thenReturn(getGoodResponse(202));
+        NameListHolder locations = new NameListHolder();
+        State state = relatedFilesRequest.complete(locations);
+        assertThat(state, is(State.COMPLETED));
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/test/java/com/connexta/alliance/nsili/endpoint/TestGetRelatedFilesRequestImpl.java]: These 4 lines seen to be repeated a lot, can we consolidate this into an `assertState` method?

File: catalog/nsili/catalog-nsili-endpoint/src/test/java/com/connexta/alliance/nsili/endpoint/TestGetRelatedFilesRequestImpl.java [hunk 13]
```diff
@@ -0,0 +1,553 @@
...
+        relatedFilesRequest.register_callback(null);
+    }
+
+    @Test
+    public void testFreeCallback() throws Exception {
+        relatedFilesRequest.free_callback("123");
+    }
+
+    @After
+    public void tearDown() {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/test/java/com/connexta/alliance/nsili/endpoint/TestGetRelatedFilesRequestImpl.java]: Remove?

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/GetParametersRequestImpl.java [hunk 14]
```diff
@@ -81,4 +144,49 @@ public RequestManager get_request_manager() throws ProcessingFault, SystemFault
...
+        Result result = null;
+        try {
+            QueryResultsCallable queryCallable = new QueryResultsCallable(queryRequest);
+            List<Result> results = guestSubject.execute(queryCallable);
+            if (results != null && !results.isEmpty()) {
+                result = results.iterator()
+                        .next();
+            }
+
+        } catch (Exception e) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/GetParametersRequestImpl.java]: Can we avoid catching generic Exception here?

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/ProductMgrImpl.java [hunk 15]
```diff
@@ -32,73 +44,123 @@
...
+            _poa().activate_object_with_id(id.getBytes(Charset.forName(NsiliEndpoint.ENCODING)),
                     getParametersRequest);
-        } catch (ServantAlreadyActive | ObjectAlreadyActive | WrongPolicy e) {
+
+            org.omg.CORBA.Object obj = _poa().create_reference_with_id(id.getBytes(Charset.forName(
+                    NsiliEndpoint.ENCODING)), GetParametersRequestHelper.id());
+            getParamRequest = GetParametersRequestHelper.narrow(obj);
+
+        } catch (WrongAdapter | WrongPolicy | UnsupportedEncodingException e) {
+            LOGGER.error("Unable to get ID from product reference: " + e);
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/ProductMgrImpl.java]: Use parameterized logging

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/NsiliEndpoint.java [hunk 16]
```diff
@@ -70,16 +77,16 @@ public NsiliEndpoint() {
...
-            orb.destroy();
-        }
-
         if (orbRunThread != null) {
             orbRunThread.interrupt();
             orbRunThread = null;
         }
 
+        if (orb != null) {
+            LOGGER.debug("Attempting to stop ORB on port: " + corbaPort);
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/NsiliEndpoint.java]: Use parameterized logging

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/AccessManagerImpl.java [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+        catalogQuery.setPageSize(10);
+
+        QueryRequestImpl catalogQueryRequest = new QueryRequestImpl(catalogQuery);
+
+        try {
+            QueryResultsCallable
+                    queryCallable = new QueryResultsCallable(catalogQueryRequest);
+            results.addAll(subject.execute(queryCallable));
+
+        } catch (Exception e) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/AccessManagerImpl.java]: Can we avoid catching generic Exception here?

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/SubmitQueryRequestImpl.java [hunk 18]
```diff
@@ -132,19 +165,85 @@ public void cancel() throws ProcessingFault, SystemFault {
...
         return new _RequestManagerStub();
     }
 
+    private void notifyCallbacks() {
+        if (!callbacks.isEmpty()) {
+            for (Callback callback : callbacks.values()) {
+                RequestDescription requestDescription = new RequestDescription();
+                try {
+                    callback._notify(State.COMPLETED, requestDescription);
+                } catch (Exception e) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/SubmitQueryRequestImpl.java]: Can we avoid catching generic Exception here?

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/SubmitQueryRequestImpl.java [hunk 19]
```diff
@@ -132,19 +165,85 @@ public void cancel() throws ProcessingFault, SystemFault {
...
+            catalogQuery.setTimeoutMillis(timeout * 1000);
+        }
+
+        QueryRequestImpl catalogQueryRequest = new QueryRequestImpl(catalogQuery);
+
+        try {
+            QueryResultsCallable queryCallable = new QueryResultsCallable(catalogQueryRequest);
+            results.addAll(guestSubject.execute(queryCallable));
+
+        } catch (Exception e) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/SubmitQueryRequestImpl.java]: Can we avoid catching generic Exception here?

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/AccessManagerImpl.java [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+            try {
+                URL url = new URL(urlStr);
+                URLConnection urlConnection = url.openConnection();
+                urlConnection.connect();
+
+                int contentLength = urlConnection.getContentLength();
+                if (contentLength > 0) {
+                    return true;
+                }
+            } catch (IOException ignore) {}
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/AccessManagerImpl.java]: Why ignore this exception?

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/GetRelatedFilesRequestImpl.java [hunk 21]
```diff
@@ -64,17 +158,54 @@ public void cancel() throws ProcessingFault, SystemFault {
...
+
+        HttpPut httpPut = new HttpPut(urlStr);
+        HttpEntity entity = new ByteArrayEntity(thumbnailBytes);
+        httpPut.setEntity(entity);
+
+        Header contentTypeHeader = new BasicHeader("Content-Type", mediaType.toString());
+        httpPut.addHeader(contentTypeHeader);
+        HttpResponse response = httpClient.execute(httpPut);
+        int statusCode = response.getStatusLine().getStatusCode();
+        if (! (statusCode == 200 || statusCode == 201 || statusCode == 202 || statusCode == 204)) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/requests/GetRelatedFilesRequestImpl.java]: Nit, but I think you can use HttpStatus.SC_OK, HttpStatus.SC_CREATED, HttpStatus.SC_ACCEPTED, and HttpStatus.SC_NO_CONTENT here

File: catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/AccessManagerImpl.java (hunk 20 above)
Comment (troymohl) [inline catalog/nsili/catalog-nsili-endpoint/src/main/java/com/connexta/alliance/nsili/endpoint/managers/AccessManagerImpl.java]: If we get an IOException when attempting to access the URL it isn't valid for the purpose of this method, so we can ignore the exception as the method defaults to returning false.

**Comment by troymohl** [comment]
@bdeining Addressed all of your comments.

**Comment by bdeining** [comment]
:+1:

**Comment by jaymcnallie** [comment]
LGTM 👍

**Pull request merged**